target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_mstc_boot.c

```diff
@@ -0,0 +1,274 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * a mtdsplit parser using "bootnum" value in the "persist" partition
+ * for the devices manufactured by MSTC (MitraStar Technology Corp.)
+ */
+#include <linux/module.h>
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/partitions.h>
+#include <linux/types.h>
+#include <linux/byteorder/generic.h>
+#include <linux/slab.h>
+#include <linux/libfdt.h>
+#include <linux/of_fdt.h>
+#include <dt-bindings/mtd/partitions/uimage.h>
+
+#include "mtdsplit.h"
+
+#define PERSIST_BOOTNUM_OFFSET	0x4
+#define NR_PARTS_MAX		2
+
+/*
+ * Legacy format image header,
+ * all data in network byte order (aka natural aka bigendian).
+ */
+ struct uimage_header {
+	uint32_t	ih_magic;	/* Image Header Magic Number	*/
+	uint32_t	ih_hcrc;	/* Image Header CRC Checksum	*/
+	uint32_t	ih_time;	/* Image Creation Timestamp	*/
+	uint32_t	ih_size;	/* Image Data Size		*/
+	uint32_t	ih_load;	/* Data	 Load  Address		*/
+	uint32_t	ih_ep;		/* Entry Point Address		*/
+	uint32_t	ih_dcrc;	/* Image Data CRC Checksum	*/
+	uint8_t		ih_os;		/* Operating System		*/
+	uint8_t		ih_arch;	/* CPU architecture		*/
+	uint8_t		ih_type;	/* Image Type			*/
+	uint8_t		ih_comp;	/* Compression Type		*/
+	uint8_t		ih_name[IH_NMLEN];	/* Image Name		*/
+};
+
+/* check whether the current mtd device is active or not */
+static int
+mstcboot_is_active(struct mtd_info *mtd, u32 *bootnum_dt)
+{
+	struct device_node *np = mtd_get_of_node(mtd);
+	struct device_node *persist_np;
+	size_t retlen;
+	u32 persist_offset;
+	u_char bootnum;
+	int ret;
+
+	ret = of_property_read_u32(np, "mstc,bootnum", bootnum_dt);
+	if (ret)
+		return ret;
+
+	persist_np = of_parse_phandle(np, "mstc,persist", 0);
+	if (!persist_np)
+		return -ENODATA;
+	/* is "persist" under the same node? */
+	if (persist_np->parent != np->parent)
+		return -EINVAL;
+	ret = of_property_read_u32(persist_np, "reg", &persist_offset);
+	of_node_put(persist_np);
+	if (ret)
+		return ret;
+	ret = mtd_read(mtd->parent, persist_offset + PERSIST_BOOTNUM_OFFSET,
+		       1, &retlen, &bootnum);
+	if (ret)
+		return ret;
+	if (retlen != 1)
+		return -EIO;
+
+	switch (bootnum) {
+	case 0 ... 2:
```

Comment:
Is the restriction for partition number 0,1 and 2 needed?
have you seen broken data here?

Reply:
I haven't seen broken data, and indeed, this check seems unnecessary. I'll drop...